zypp/zyppng/media/medianetwork.cc

```diff
@@ -214,49 +516,161 @@ void MediaHandlerNetwork::handleRequestResult( const Download &req ) const
         ZYPP_THROW( zypp::media::MediaTimeoutException(reqUrl) );
         break;
       case NetworkRequestError::Forbidden:
-        ZYPP_THROW( zypp::media::MediaForbiddenException(reqUrl, err.toString()));
+        ZYPP_THROW( zypp::media::MediaForbiddenException(reqUrl, err.errordesc()));
         break;
       case NetworkRequestError::NotFound:
-        ZYPP_THROW( zypp::media::MediaFileNotFoundException(reqUrl, req.targetPath()) );
+        ZYPP_THROW( zypp::media::MediaFileNotFoundException( reqUrl, filename ) );
         break;
       default:
         break;
     }
-    ZYPP_THROW( zypp::media::MediaCurlException( reqUrl, err.toString(), err.nativeErrorString() ) );
+    ZYPP_THROW( zypp::media::MediaCurlException( reqUrl, err.errordesc(), err.nativeerror() ) );
   }
-
-  ZYPP_THROW( zypp::media::MediaCurlException( req.url(), req.errorString(), "" ) );
+  ZYPP_THROW( zypp::media::MediaCurlException( req.url(), err.errordesc() , "" ) );
 }
 
-void MediaHandlerNetwork::authenticate( const Download &, NetworkAuthData &auth, const std::string &availAuth ) const
+bool MediaHandlerNetwork::retry( DispatchContext &ctx, Request &req ) const
 {
+  const auto &res = req.result();
+  if ( !res || !res->has_error() )
+    return false;
+
+  const auto &resErr = res->error();
+  if ( resErr.error() != NetworkRequestError::AuthFailed && resErr.error() != NetworkRequestError::Unauthorized )
+    return false;
+
+  const std::string &availAuth =  resErr.extra_info().count("authHint") ? resErr.extra_info().at("authHint") : "";
 
   zypp::callback::SendReport<zypp::media::AuthenticationReport> auth_report;
 
-  NetworkAuthData authCopy = auth;
+  zypp::media::CredentialManager cm( zypp::media::CredManagerOptions( zypp::ZConfig::instance().repoManagerRoot()) );
+  auto authDataPtr = cm.getCred( _url );
+
+  NetworkAuthData_Ptr curlcred( new NetworkAuthData() );
+  curlcred->setUrl( _url );
 
   // preset the username if present in current url
-  if ( !_url.getUsername().empty() && authCopy.password().empty() )
-    authCopy.setUsername( _url.getUsername() );
+  if ( !_url.getUsername().empty() )
+    curlcred->setUsername(_url.getUsername());
+  // if CM has found some credentials, preset the username from there
+  else if ( authDataPtr )
+    curlcred->setUsername( authDataPtr->username() );
+
+  // set available authentication types
+  // might be needed in prompt
+  curlcred->setAuthType( availAuth );
 
   std::string prompt_msg = zypp::str::Format( _("Authentication required for '%s'") ) % _url.asString();
 
-  // set available authentication types from the exception
-  // might be needed in prompt
-  authCopy.setAuthType(availAuth);
+  // the auth data that was used in the request is older than whats in the Credential manager,
+  // so instead of asking the user for creds we first try to use the new ones
+  if ( res->last_auth_timestamp() > 0 && res->last_auth_timestamp() < authDataPtr->lastDatabaseUpdate() ) {
```

Comment:
```
/home/ma/zypp/libzypp/zypp/zyppng/media/medianetwork.cc:567:69: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if ( res->last_auth_timestamp() > 0 && res->last_auth_timestamp() < authDataPtr->lastDatabaseUpdate() ) {
```
`last_auth_timestamp` should match `time_t`

Reply:
Done